server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/filemove/FileMoveDetectionStep.java

```diff
@@ -182,64 +195,133 @@ public void visitFile(Component file) {
+    @Override
+    public void handleResult(ResultContext<? extends LineHashesWithKeyDto> resultContext) {
+      LineHashesWithKeyDto lineHashesDto = resultContext.getResultObject();
+      if (lineHashesDto.getPath() == null) {
+        return;
+      }
+      int removeFileIndex = removedFilesIndexes.get(lineHashesDto.getKey());
+      ScoreMatrix.ScoreFile removedFile = removedFiles[removeFileIndex];
+      int lowerBound = (int) Math.floor(removedFile.getLineCount() * LOWER_BOUND_RATIO);
+      int upperBound = (int) Math.ceil(removedFile.getLineCount() * UPPER_BOUND_RATIO);
+
+      for (int newFileIndex = 0; newFileIndex < newFiles.length; newFileIndex++) {
+        ScoreMatrix.ScoreFile newFile = newFiles[newFileIndex];
+        if (newFile.getLineCount() >= upperBound) {
           continue;
         }
+        if (newFile.getLineCount() <= lowerBound) {
+          break;
+        }
 
-        int reportFileIndex = 0;
-        for (Map.Entry<String, File> reportFileSourceAndKey : reportFileSourcesByKey.entrySet()) {
-          File unmatchedFile = reportFileSourceAndKey.getValue();
-          int score = fileSimilarity.score(fileInDb, unmatchedFile);
-          scoreMatrix[dbFileIndex][reportFileIndex] = score;
-          if (score > maxScore) {
-            maxScore = score;
-          }
-          reportFileIndex++;
+        File fileInDb = new FileImpl(lineHashesDto.getPath(), lineHashesDto.getLineHashes());
+        File unmatchedFile = newFileSourcesByKey.get(newFile.getFileKey());
+        int score = fileSimilarity.score(fileInDb, unmatchedFile);
+        scoreMatrix[removeFileIndex][newFileIndex] = score;
+        if (score > maxScore) {
+          maxScore = score;
         }
-        dbFileIndex++;
       }
     }
 
-    return new ScoreMatrix(dbFileKeys, reportFileSourcesByKey, scoreMatrix, maxScore);
-  }
-
-  @CheckForNull
-  private File getFile(DbSession dbSession, DbComponent dbComponent) {
-    if (dbComponent.getPath() == null) {
-      return null;
-    }
-    FileSourceDto fileSourceDto = dbClient.fileSourceDao().selectSourceByFileUuid(dbSession, dbComponent.getUuid());
-    if (fileSourceDto == null) {
-      return null;
+    int getMaxScore() {
+      return maxScore;
     }
-    String lineHashes = firstNonNull(fileSourceDto.getLineHashes(), "");
-    return new File(dbComponent.getPath(), LINES_HASHES_SPLITTER.splitToList(lineHashes));
   }
 
   private static void printIfDebug(ScoreMatrix scoreMatrix) {
     if (LOG.isDebugEnabled()) {
-      LOG.debug("ScoreMatrix:\n" + scoreMatrix.toCsv(';'));
+      try {
+        Path tempFile = Files.createTempFile("score-matrix", ".csv");
+        try (BufferedWriter writer = Files.newBufferedWriter(tempFile, Charset.forName("UTF-8"))) {
+          writer.write(scoreMatrix.toCsv(';'));
+        }
+        LOG.info("score matrix dumped as CSV: {}", tempFile);
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
     }
+    // if (LOG.isDebugEnabled()) {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: This block of commented-out lines of code should be removed. ([squid:CommentedOutCodeLine](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&pullRequest=3066&issues=AWIvtD1FdSOV8npVrlzy&open=AWIvtD1FdSOV8npVrlzy)

[ISSUE_KEY]: AWIvtD1FdSOV8npVrlzy